requester/requester.go

```diff
@@ -40,20 +49,41 @@ func New(cfg config.Config) *Requester {
 
 // Run starts the benchmark test and pipelines the results inside a Report.
 // Returns the Report when the test ended and all results have been collected.
-func (r *Requester) Run() Report {
+func (r *Requester) Run() (Report, error) {
+	req, err := r.config.HTTPRequest()
+	if err != nil {
+		return Report{}, fmt.Errorf("%w: %s", ErrRequest, err)
+	}
+
+	if err := r.ping(req); err != nil {
+		return Report{}, fmt.Errorf("%w: %s", ErrConnection, err)
+	}
+
 	ctx, cancel := context.WithTimeout(context.Background(), r.config.RunnerOptions.GlobalTimeout)
+	errCh := make(chan error)
 
 	go func() {
 		defer cancel()
 		defer close(r.records)
-		semimpl.Do(ctx,
-			r.config.RunnerOptions.Concurrency,
-			r.config.RunnerOptions.Requests,
-			r.record,
-		)
+
+		errCh <- dispatcher.
+			New(r.config.RunnerOptions.Concurrency).
+			Do(ctx, r.config.RunnerOptions.Requests, r.record(req))
 	}()
 
-	return r.collect()
+	if err := <-errCh; err != nil {
+		return Report{}, err
+	}
+
+	return r.collect(), nil
```

Comment:
I might be wrong but:
I'm pretty sure reading from `errCh` will block until something from `requester.Do` writes to it.

Meaning `r.collect()` is not run in parallel of `Do`. Losing the whole point.

---

Suggestion/path to explore, untested:

```suggestion
	return r.collect(ctx, errCh)
```
And this `r.collect`:
```go
func (r *Requester) collect(ctx context.Context, errc <-chan error) (Report, error) {
	rep := Report{
		Config:  r.config,
		Records: make([]Record, 0, r.config.RunnerOptions.Requests), // Provide capacity if known.
	}
	for {
		select {
		case <-ctx.Done():
			return rep, nil
		case err := <-errc:
			return Report{}, err
		case rec := <-r.records:
			if rec.Error != nil {
				rep.Fail++
			} else {
				rep.Success++
			}
			rep.Records = append(rep.Records, rec)
			rep.Length++
		}
	}
}
```

Note: we go back to incremental Report building.

Maybe going back to safe slice and mutex is better.

Reply:
Oh nice catch indeed. Will fix it as you suggest, and propose a second PR with mutexes targeting this one so we can compare.